streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java

```
@@ -502,6 +517,13 @@ public StreamsConfig(final Map<?, ?> props) {
         consumerProps.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, originals().get(BOOTSTRAP_SERVERS_CONFIG));
         // remove deprecate ZK config
         consumerProps.remove(ZOOKEEPER_CONNECT_CONFIG);
+        if (eosEnabled) {
+            if (clientProvidedProps.containsKey(ConsumerConfig.ISOLATION_LEVEL_CONFIG)) {
```

Why throw an exception? Can't we just log it and remove it from the config.

---

Because this is current default behavior. I agree we can log and we do have a JIRA and PR for this already (https://issues.apache.org/jira/browse/KAFKA-5096). Did not update this code yet accordingly...

---

What is the benefit of logging a warning instead of throwing an exception? If the user explicitly specified a config and we override it, it makes sense to throw an exception so that the user tests show the issue (i.e. fail fast). The user can then figure out if the config was a mistake or if they misunderstood some aspect and need an alternative solution.

This is what the other clients (e.g. producer) do.

---

The point is, that the user cannot set this config, and there is only _one_ valid value. If there would be multiple valid values, it would make sense to throw (as we should not "randomly" pick one) -- but as there is only one, we can just set it an run (and log). This gives a better out-of-the-box experience IMHO -- it's more "resilient" and "self-healing" if you wish (IMHO, less exceptions are always better if the exceptions are not really needed as there is only one fix)

---

It would be nice if we had consistent behaviour across clients unless there's a good reason to deviate. Here's a very similar example in the producer where we throw an exception (fail-fast):

https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java#L394

Maybe you can sync with @apurvam and @hachikuji on this. Either way, it should not block this PR as it can be tweaked in a follow-up along with other changes that are planned.

---

IMHO, Streams is more than a client. Thus, it seems ok to behave differently. \cc @guozhangwang @dguy @enothereska @miguno @sriramsub

---

Yes - my opinion in this case is to log it and move on. I don't see it as exceptional, but whatever. I'm happy either way

---

I would suggest we throw an exception here: this is the case that users are not allowed to specify these configs (e.g. this and `if (clientProvidedProps.containsKey(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG))`), if they do specify, we should fail fast to notify them clearly.

---

We had a discussion about this, and agreed to not throw but overwrite and log. We also created a JIRA for this change: https://issues.apache.org/jira/browse/KAFKA-5096 (with an open PR).

We can discuss again of course. Waiting for further feedback.